core/lib/zksync_core/src/sync_layer/external_io.rs

```diff
@@ -141,6 +141,7 @@ impl StateKeeperIO for ExternalIO {
             cursor.next_l2_block,
         );
 
+        clear_pending_l2_block(&mut storage, cursor.next_l2_block - 1).await?;
```

Comment:
Could this be placed in `StateKeeperPersistence` (namely, `StateKeeperOutputHandler::initialize()` impl)? If so, it would be more local (= easier to reason about) and wouldn't require adding to both `StateKeeperIO` impls and sort of breaking abstraction boundaries.

Reply:
So, I put it into IO::initialize because it's initialized first.
```rust
async fn run_inner(&mut self) -> Result<Infallible, Error> {
        let (cursor, pending_batch_params) = self.io.initialize().await?;
        self.output_handler.initialize(&cursor).await?;
        ...
```
Both IOs and StateKeeperPersistence in the current form require an absence of pending batch to work properly. We can change but that means adapting more SQL queries so I think current approach is better.
It doesn't look like a good solution for `StateKeeperOutputHandler` to rely on the fact that 1) IO is initialized first 2) IO cleans pending block. Maybe it's worth to run `clear_pending_l2_block` in both places, WDYT?

Reply:
I see. IMO, adapting SQL queries is not worth the effort; I'm fine with the `clear_pending_l2_block()` in the output handler (or ensuring that a pending L2 block doesn't exist) in the short term. Abstractly, this issue makes me wonder whether distribution of responsibilities between `StateKeeperIO` and `StateKeeperOutputHandler` is chosen correctly; maybe, it would make sense to initialize the output handler first and make it return `IoCursor` that would then be provided to `StateKeeperIO`?

Reply:
hmm, I'm not sure about distribution. IMO it's ok to set one of these abstractions to be responsible for clean-up and initialize it first + document. For now will do clean-up in both